src/bootstrap/ci_profiler.rs

```diff
@@ -0,0 +1,216 @@
+use crate::Build;
+use build_helper::t;
+use serde::Serialize;
+use std::fs::File;
+use std::sync::mpsc::{Receiver, RecvTimeoutError, Sender};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+use sysinfo::{ProcessorExt, System, SystemExt};
+
+const CPU_SAMPLING_INTERVAL: Duration = Duration::from_secs(1);
+
+#[derive(Clone)]
+pub(crate) struct CiProfiler {
+    state: Arc<Mutex<ProfilerState>>,
+}
+
+impl CiProfiler {
+    pub(crate) fn init() -> Self {
+        let (reset_send, reset_recv) = std::sync::mpsc::channel();
+
+        let state = Arc::new(Mutex::new(ProfilerState {
+            finished_steps: Vec::new(),
+            running_steps: Vec::new(),
+
+            cpu_single_core: Vec::new(),
+            cpu_all_cores: Vec::new(),
+            timer_start: None,
+
+            reset_cpu_check: reset_send,
+        }));
+
+        let state_cloned = state.clone();
+        std::thread::spawn(move || cpu_sampler_thread(state_cloned, reset_recv));
+
+        CiProfiler { state }
+    }
+
+    pub(crate) fn enter_step<S: Step>(&self, step: &S) {
+        let mut state = t!(self.state.lock());
+
+        // Consider all the stats gathered so far as the parent's.
+        if !state.running_steps.is_empty() {
+            self.collect_stats(&mut *state);
+        }
+
+        let _ = state.reset_cpu_check.send(());
+
+        state.cpu_single_core.clear();
+        state.cpu_all_cores.clear();
+        state.timer_start = Some(Instant::now());
+
+        state.running_steps.push(ProfiledStep {
+            type_: std::any::type_name::<S>().into(),
+            debug_repr: format!("{step:?}"),
+
+            duration: Duration::ZERO,
+            cpu_single_core: Vec::new(),
+            cpu_all_cores: Vec::new(),
+
+            childs: Vec::new(),
+        });
+    }
+
+    pub(crate) fn exit_step(&self) {
+        let mut state = t!(self.state.lock());
+
+        self.collect_stats(&mut *state);
+
+        let step = state.running_steps.pop().unwrap();
+        if state.running_steps.is_empty() {
+            state.finished_steps.push(step);
+            state.timer_start = None;
+        } else {
+            state.running_steps.last_mut().unwrap().childs.push(step);
+
+            // Resume the timer for the parent step.
+            state.timer_start = Some(Instant::now());
+        }
+    }
+
+    fn collect_stats(&self, state: &mut ProfilerState) {
+        let step = state.running_steps.last_mut().unwrap();
+
+        step.duration += state.timer_start.unwrap().elapsed();
+        step.cpu_single_core.append(&mut state.cpu_single_core);
+        step.cpu_all_cores.append(&mut state.cpu_all_cores);
+    }
+
+    pub(crate) fn persist(&self, build: &Build) {
+        let mut state = t!(self.state.lock());
+        assert!(state.running_steps.is_empty(), "steps are still executing");
+
+        let steps = std::mem::take(&mut state.finished_steps);
+        let json = vec![JsonNode::RustbuildInvocation {
```

Comment:
IMO we should define the top-level structure of this file to be extensible (i.e., an object, rather than an array).

Reply:
I agree the format should be extensible, but I think the schema I have achieves most of that goal.

As discussed in other comments, we want to track multiple `x.py` invocations into a single file. But, if we had a top-level object, the object wouldn't be able to store information that differs between invocations, such as the wall clock time of the whole invocation.

In the current schema, the role of the top-level object is in the `RustbuildInvocation` node, and all the metadata would be stored there. Having the top-level array then allows to easily store multiple of those invocations into a single file.

The only metadata I can see us wanting to store in the top level is maybe a format version number, but I'm not sure it's worth changing the schema just for that. What do you think?

Reply:
At least some consumers -- e.g., total wall time analysis like my rustc-ci-timing scripts -- don't care about individual rustbuild invocations.

I also think in general a top-level object future-proofs us against any expansion we want to do, and in general unless there's a strong motivation to avoid it I'd prefer to have it (even if the only field is a "invocations" array).

As I mentioned elsewhere there are some properties that are more machine-dependent than invocation-specific that I'd like to add in the future (e.g., CPU model, disk space, etc.) that are useful to have on a per-job level IMO. Those are not really well-mapped to an individual x.py invocation. Likewise, I think it is useful to have the ability to represent e.g. CPU time spent *outside* x.py invocations, for example if we start trying to track the duration of AWS uploads or docker builds. Both of those don't map well to "RustbuildInvocations", and while we can have an array of differently-typed objects, that gets somewhat unwieldy fast, IMO.